Updating contribution guide (#1209)
opentofu/opentofu: CONTRIBUTING.md

Welcome and thank you for wanting to contribute! 

> ## Get started

Reviewer: How about mentioning Weekly Updates here? Something along the lines of:

- Want to get the latest updates? Check the [WEEKLY_UPDATES.md](https://github.com/opentofu/opentofu/blob/main/WEEKLY_UPDATES.md) 

Bonus points if you folks want to commit to the weekly DEV gatherings on Wednesdays and mention it here with the calendar invite 😅 and maybe mention the TSC (which is mentioned at the end of the document).

Author: Good point. If the Wednesday meetings become regular, we could add them here.

  3. Wait for a maintainer to assign it to you,
  4. Then [submit your code here ➡️](https://github.com/opentofu/opentofu/compare)
- Want to fix a bug? [Submit a PR here ➡️](https://github.com/opentofu/opentofu/compare)
- Want to know what's going on? Read the [weekly updates ➡️](WEEKLY_UPDATES.md), the [TSC summary ➡️](TSC_SUMMARY.md) or join the [community meetings ➡️](https://meet.google.com/xfm-cgms-has) on Wednesdays at 14:30 CET / 8:30 AM Eastern / 5:30 AM Western / 19:00 India time on this link: https://meet.google.com/xfm-cgms-has ([📅 calendar link](https://calendar.google.com/calendar/event?eid=NDg0aWl2Y3U1aHFva3N0bGhyMHBhNzdpZmsgY18zZjJkZDNjMWZlMGVmNGU5M2VmM2ZjNDU2Y2EyZGQyMTlhMmU4ZmQ4NWY2YjQwNzUwYWYxNmMzZGYzNzBiZjkzQGc))

> [!TIP]

> > For more OpenTofu events, subscribe to the [OpenTofu Events Calendar](https://calendar.google.com/calendar/embed?src=c_3f2dd3c1fe0ef4e93ef3fc456ca2dd219a2e8fd85f6b40750af16c3df370bf93%40group.calendar.google.com)!

Reviewer 2: 🥳

### Building OpenTofu

There are two ways to build OpenTofu: natively and in a container. Building natively means you are building directly where your code is located without dynamically creating a container. In contrast, building in a container means you are using a build container to run the build process only.

#### Building natively

To build OpenTofu natively, you will need to install Go in the environment you are running in. You can then run the `go build` command from your OpenTofu source directory as follows:

```sh
GOOS=linux GOARCH=amd64 go build -o tofu -v -buildvcs=false ./cmd/tofu
```

This command will produce a `tofu` binary in your current directory, which you can test by running `./tofu --version`.

> [!TIP]
> Replace the `GOOS` and `GOARCH` values with your target platform if you wish to cross-compile.

#### Building in a container

If you have Docker or Podman installed, you can run the entire build process in a container too:

```sh
docker run \
  --rm \
  -v "$PWD":/usr/src/opentofu\
  -w /usr/src/opentofu golang:1.21.3\
  GOOS=linux GOARCH=amd64 go build -o tofu -v -buildvcs=false ./cmd/tofu
```

> This will create the `opentofu` binary in the current working directory, which you can test by running `./tofu --version`.

Reviewer 3: `tofu`

- Fix linter warnings for tools not currently in the CI pipeline
- Add new CI tooling

---

## Writing code for OpenTofu

Eager to get started on coding? Here's the short version:

1. Set up a Go development environment with Git.

> 2. Pay attention to copyright: write the code yourself, avoid copy/paste. Disable or limit your AI coding assistant.

Reviewer 4: it may be worth linking to the DCO here for more information too.

Reviewer 2: I agree

3. Run the tests with `go test` in the package you are working on.
4. Build OpenTofu by running `go build ./cmd/tofu`.
5. Update [the changelog](CHANGELOG.md).
6. When you commit, use `git commit -s` to sign off your commits.

---

### Setting up your development environment

> You can develop OpenTofu on any platform you like. However, some tests are currently failing on Windows, so we recommend either a Linux (WSL works) or MacOS build environment. You will need to install [Go](https://golang.org/) and [Git](https://git-scm.com/) installed, and we recommend an IDE to help you with code completion and code quality warnings. (We recommend installing the Go version documented in the [.go-version](.go-version) file.)

Reviewer 4: ```suggestion
You can develop OpenTofu on any platform you like. However, we recommend either a Linux (including WSL on Windows) or a MacOS build environment. You will need [Go](https://golang.org/) and [Git](https://git-scm.com/) installed, and we recommend an IDE to help you with code completion and code quality warnings. (We recommend installing the Go version documented in the [.go-version](.go-version) file.)
```

> If you use Visual Studio Code or Goland/IntelliJ and have Docker or Podman installed, you can also use a [devcontainer](.devcontainer.json) instead of installing the tools directly on your machine. In Visual Studio Code, you can install the [Remote Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers), then reopen the project to get a prompt about activating the devcontainer. In Goland/Intellij, open the `.devcontainers.json` file and click the purple cube icon that appears next to the line numbers to activate the dev container. At this point you can proceed as if you were [building natively](#building-natively) on Linux.

Reviewer 4: ```suggestion
Alternatively, if you use Visual Studio Code or Goland/IntelliJ and have Docker or Podman installed, you can also use a [devcontainer](.devcontainer.json). In Visual Studio Code, you can install the [Remote Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers), then reopen the project to get a prompt about activating the devcontainer. In Goland/Intellij, open the `.devcontainers.json` file and click the purple cube icon that appears next to the line numbers to activate the dev container. At this point you can proceed as if you were [building natively](#building-natively) on Linux.
```

---

### Building OpenTofu

There are two ways to build OpenTofu: natively and in a container. Building natively means you are building directly where your code is located without dynamically creating a container. In contrast, building in a container means you are using a build container to run the build process only.

#### Building natively

To build OpenTofu natively, you will need to install Go in the environment you are running in. You can then run the `go build` command from your OpenTofu source directory as follows:

```sh
GOOS=linux GOARCH=amd64 go build -o tofu -v -buildvcs=false ./cmd/tofu
```

This command will produce a `tofu` binary in your current directory, which you can test by running `./tofu --version`.

> [!TIP]

> > Replace the `GOOS` and `GOARCH` values with your target platform if you wish to cross-compile.

Reviewer 4: might be a good idea to link to the docs on this (https://pkg.go.dev/cmd/go#hdr-Compile_and_run_Go_program)

#### Building in a container

> If you have Docker or Podman installed, you can run the entire build process in a container too:

Reviewer 4: ```suggestion
If you have Docker or a compatible alternative installed, you can run the entire build process in a container too:
```

Lots of people love tools like orbstack, cri/containerd etc, there's lots of alternatives too.

```sh
docker run \
  --rm \
  -v "$PWD":/usr/src/opentofu\
  -w /usr/src/opentofu golang:1.21.3\
  GOOS=linux GOARCH=amd64 go build -o tofu -v -buildvcs=false ./cmd/tofu
```

This will create the `tofu` binary in the current working directory, which you can test by running `./tofu --version`.

> > [!TIP]

Reviewer 4: Do we need this tip in here twice?

> Replace the `GOOS` and `GOARCH` values with your target platform if you wish to cross-compile. 

---

### Running tests

Similar to builds, you can use the `go test` command to run tests. To run the entire test suite, please run the following command in your OpenTofu source directory:

```sh
go test ./...
```

> Alternatively, you can also run the `go test` command in the package you are currently working on.

Reviewer 4: maybe a link to the docs for more ways people can test (https://pkg.go.dev/cmd/go#hdr-Test_packages) ?

---

### Updating the changelog

We are keeping track of the changes to OpenTofu in the [CHANGELOG.md](CHANGELOG.md) file. Please update it when you add features or fix bugs in OpeTofu.

---

### Signing off your commits

When you contribute code to OpenTofu, we require you to add a [Developer Certificate of Origin](https://developercertificate.org/) sign-off. Please read the DCO carefully before you proceed and only contribute code you have written yourself. Please do not add code that you have not written from scratch yourself without discussing it in the related issue first.

The simplest way to add a sign-off is to use the `-s` command when you commit:

```
git commit -s -m "My commit message"
```

> [!IMPORTANT]
> Make sure your `user.name` and `user.email` setting in Git matches your GitHub settings. This will allow the automated DCO check to pass and avoid delays when merging your PR.

> [!TIP]

> > Have you forgotten your sign-off? You can use `git rebase --signoff` to add it after the fact.

Reviewer 5: I think you can also sign the DCO manually, without any Git shenanigans?

Author: Yes, if you make your contribution on the GitHub web interface. Otherwise you'll have to do git shenanigans, unfortunately.

Reviewer 5: I'm not 100% sure. I always forget to sign off, push my code via the CLI and then just sign the DCO manually from the link on the red check.

Author: Interesting, I'll have to check if that's available to external contributors.

Author: Checked, unfortunately that option is only available to maintainers as an override.

Reviewer 5: Ah, OK, good to know.

> 

Reviewer 4: It may be worth mentioning that this works as a check on a pull request, and if it fails they can check the details of the failing check for more information about why.

---

### A note on copyright

We take copyright and intellectual property very seriously. A few quick rules should help you:

> 1. You should write most, if not all of your code yourself.

Reviewer 5: Can we start with the note about LLMs, and then put this point in the context of that? Otherwise it's a bit unclear - does it mean I can't submit my friend's code? Or does it refer to not copying from the legacy repo?

Author: Technically no, you shouldn't submit your friends code without their permission. This may need rephrasing.

Author: Rephrased.

2. Do not copy code from other authors without permission. If you have permission, always add the `Co-authored-by` sign-off to your commits.
3. When you copy/paste code from within the OpenTofu code, always make it explicit where you copied from.
4. When you copy from external sources, always make explicit where your copy is coming from and make sure that their license and copyright attribution requirements are met.

> 5. Specifically, do not copy from the Terraform repository, or any PRs filed against that repository. Doing so will immediately disqualify your PR from being merged.

Reviewer 5: > Doing so will immediately disqualify your PR from being merged.

...and prevent you from submitting further contributions.

Author: Rephrased, maybe this works better?

> [!WARNING]

> > Coding assistants based on large language models (LLMs), such as ChatGPT or GitHub Copilot, are awesome tools to help. However, in the specific case of OpenTofu the training data may include the BSL-licensed Terraform. Since the OpenTofu/Terraform codebase is very specific and LLMs don't have any other training source, they may emit copyrighted code. Please avoid using LLM-based coding assistants as much as possible. If your PR is found to contain code matching Terraform, your PR will be rejected, and you will not be permitted to work on that feature.

Reviewer 4: @Evi1Pumpkin you've been writing some stuff up in this area, would you mind looking at this please?

Author: I've slightly updated this section but would still like some feedback @Evi1Pumpkin

Reviewer 2: LGTM

Author: Thank you!

### Building Natively
---

Refer to the file [`.go-version`](.go-version) to see which version of Go OpenTofu is currently built with. Other versions will often work, but if you run into any build or testing problems please try with the specific Go version indicated. You can optionally simplify the installation of multiple specific versions of Go on your system by installing [`goenv`](https://github.com/syndbg/goenv), which reads `.go-version` and automatically selects the correct Go version.
## Advanced topics

### Build with Go
### Acceptance Tests: Testing interactions with external services

Switch into the root directory of the cloned repository and build OpenTofu using the Go toolchain in the standard way:
The test command above only runs the self-contained tests that run without external services. There are, however, some optional tests in the OpenTofu CLI codebase that *do* interact with external services. We collectively refer to them as "acceptance tests".

```sh
cd opentofu

> go install ./cmd/tofu

Reviewer 2: I wonder why you decided to change this explanation and use the `go build` instead? What was the reason?

Author: Because it's not in line with the containerized description below and it's also not intuitive that this will build and install the binaries.

- Fix linter warnings for tools not currently in the CI pipeline
- Add new CI tooling

---

## Writing code for OpenTofu

Eager to get started on coding? Here's the short version:

1. Set up a Go development environment with Git.
2. Pay attention to copyright: [please read the DCO](https://developercertificate.org/), write the code yourself, avoid copy/paste. Disable or limit your AI coding assistant.
3. Run the tests with `go test` in the package you are working on.
4. Build OpenTofu by running `go build ./cmd/tofu`.
5. Update [the changelog](CHANGELOG.md).
6. When you commit, use `git commit -s` to sign off your commits.

---

### Setting up your development environment

You can develop OpenTofu on any platform you like. However, we recommend either a Linux (including WSL on Windows) or a MacOS build environment. You will need [Go](https://golang.org/) and [Git](https://git-scm.com/) installed, and we recommend an IDE to help you with code completion and code quality warnings. (We recommend installing the Go version documented in the [.go-version](.go-version) file.)

> Alternatively, if you use Visual Studio Code or Goland/IntelliJ and have Docker or Podman installed, you can also use a [devcontainer](.devcontainer.json). In Visual Studio Code, you can install the [Remote Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers), then reopen the project to get a prompt about activating the devcontainer. In Goland/Intellij, open the `.devcontainers.json` file and click the purple cube icon that appears next to the line numbers to activate the dev container. At this point you can proceed as if you were [building natively](#building-natively) on Linux.

Reviewer: dev containers look interesting, I didn't try them, I'll check them. 

The dev containers are not directly related to how to debug the code, but I think the information in [DEBUGGING.md](https://github.com/opentofu/opentofu/blob/main/DEBUGGING.md) could be melded into this Contribution guide, or it can be mentioned there. I don't have a specific suggestion, feel free to skip this comment :)

---

> ### Building OpenTofu

Reviewer: Similar gut feeling about the https://github.com/opentofu/opentofu/blob/main/BUILDING.md . It would make sense to meld it into here, or mention the other document here, and make sure that both of the documents are aligned.

> [!TIP]
> Replace the `GOOS` and `GOARCH` values with your target platform if you wish to cross-compile. You can find more information in the [Go documentation](https://pkg.go.dev/cmd/go#hdr-Compile_and_run_Go_program).

#### Building in a container

If you have Docker or a compatible alternative installed, you can run the entire build process in a container too:

```sh
docker run \
  --rm \
  -v "$PWD":/usr/src/opentofu\
  -w /usr/src/opentofu golang:1.21.3\
  GOOS=linux GOARCH=amd64 go build -o tofu -v -buildvcs=false ./cmd/tofu
```

This will create the `tofu` binary in the current working directory, which you can test by running `./tofu --version`.

---

### Running tests

Similar to builds, you can use the `go test` command to run tests. To run the entire test suite, please run the following command in your OpenTofu source directory:

```sh
go test ./...
```

Alternatively, you can also run the `go test` command in the package you are currently working on.

> [!TIP]
> You can find more information on testing in the [Go documentation](https://pkg.go.dev/cmd/go#hdr-Test_packages).

> 

Reviewer: Why not keeping the examples for testing the packages here? 
```
go test ./internal/command/...
go test ./internal/addrs
```
I mentioned it in the old section of the PR.

4. When you copy/paste code from within the OpenTofu code, always make it explicit where you copied from. This helps us resolve issues later on.
5. Before you copy code from external sources, make sure that the license allows this. Also make sure that any licensing requirements, such as attribution, are met. When in doubt, ask first!
6. Specifically, do not copy from the Terraform repository, or any PRs others have filed against that repository. This code is licensed under the BSL, a license which is not compatible with OpenTofu. (You may submit the same PR to both Terraform and OpenTofu as long as you are the author of both.)

> [!WARNING]
> To protect the OpenTofu project from legal issues violating these rules will immediately disqualify your PR from being merged and you from working on that area of the OpenTofu code base in the future. Repeat violations may get you barred from contributing to OpenTofu. 

---

> ## Advanced topics

Reviewer: Maybe this one deserves another document, like CONTRIBUTING-ADVANCED.md or something like that?

Reviewer: or DEVELOPER-GUIDE.md?

Author: We could write a much larger docs, but my experience from [this project](https://containerssh.io/v0.5/development/devenv/) is that it will be read rarely. I think it should best be kept in one place and as short as possible.

```
TF_ACC=1 go test ./internal/initwd
```

---

### Integration Tests: Testing interactions with external backends

OpenTofu supports various [backends](https://opentofu.org/docs/language/settings/backends/configuration). We run integration test against them to ensure no side effects when using OpenTofu.

Execute to list all available commands to run tests:

```commandline
make list-integration-tests
```

From the list of output commands, you can execute those which involve backends you intend to test against.

For example, execute the command to run integration tests with s3 backend:

```commandline
make test-s3
```

---

### Generated Code

Some files in the OpenTofu CLI codebase are generated. In most cases, we update these using `go generate`, which is the standard way to encapsulate code generation steps in a Go codebase.

> ```

Reviewer: I think it might be helpful to keep those examples for testing the packages.

Note: you need to define the `GITHUB_TOKEN` environment variable to a valid GitHub personal access token, or you will hit rate limiting from the GitHub API which `licensei` uses to discover the licenses of dependencies.

---

### Backporting

All changes to OpenTofu should, by default, go into the `main` branch. When a fix is important enough, it will be backported to a version branch and release when the next minor release rolls around.

Start the backporting process by making sure both the `main` and the target version branch are up-to-date in your working copy. Look up the commit ID of the commit and then switch to your target version branch. Now create a new branch from that version branch:

```sh
git checkout -b backports/ISSUE_NUMBER
```

Now you can cherry-pick the commit in question to your backport branch:

```sh
git cherry-pick -s COMMIT_ID_HERE
```

Finally, create an additional commit to edit the changelog and send in your PR. 

---

> ## FAQ

Reviewer: I think this FAQ definitely needs a separate document of its own. 
How about moving this part to a separate document and linking it here? 

I think when the Contribution document gets bigger and longer, people might refrain reading it.

Reviewer: Oh, also, TOC please for FAQ's, or cascading questions to prevent doom scrolling through FAQ document 😅

Author: I'm a bit split on ToCs, GitHub already has them auto-generated for markdown files, but their usability is... less than ideal.

### Who decides if a feature will be implemented and how is that decision made?

When you submit an enhancement request, the [core team](#who-is-the-core-team) looks at your issue first. Given the size of the code, adding a new feature is always a careful balancing act. The core team takes the following points into consideration:

1. **Is it possible to implement this on a technical level?**<br />Sometimes, even if a feature would be extremely useful, the state of the codebase doesn't let us do it. 
2. **Does the feature cause more technical debt?**<br />A feature request may hide a larger issue under the hood. Sometimes it is more desirable to resolve the underlying issue instead of implementing the feature in isolation.
3. **Is there someone who would do the work?**<br />The core team doesn't have the capacity to implement everything, so for many issues community contributions are very welcome. Sometimes companies external to OpenTofu decide to dedicate developers for the development of a specific larger feature, which can also weigh in on the decision-making process.

> 4. **Is there enough capacity on the core team to support a community contributor?**<br />We don't expect contributors to implement the entire feature in isolation, we actively participate with planning, reviews and writing documentation as needed.

Reviewer 6: I would clarify in the description here that:

"We don't expect contributors to implement the entire feature in isolation, we actively participate with planning, reviews and writing documentation as needed. Thus, core engineers need to be available in order for an issue to be tackled, even if it's being tackled by the community."

Rephrase as needed, I don't think my version reads too nice.

3. **More information is needed.** The core team will either add questions in comments, or when there is a deep technical issue to be resolved, call for an RFC to detail a possible implementation.
4. **More community input is needed.** When an issue is, on its surface, valuable, but there is no track record of a large portion of the community needing it, the core team adds the `needs community input` label. If you are interested in the feature and would like to use it, please add a reaction to the issue and add a description on specifically what problem it would solve for you in a comment.
5. **The feature is rejected.** If based on the criteria above it is not feasible to implement the feature, the core team closes the issue with an explanation why it is being closed.
6. **The feature is referred to the Technical Steering Committee**. If the feature requires the commitment of a larger amount of core developer time, has legal implications, or otherwise requires leadership attention, the core team adds the feature to the agenda of the Technical Leadership Committee. Once decided, the TSC records the decision in the [TSC_SUMMARY.md](TSC_SUMMARY.md) file.

---

> ### What do the labels mean?

Reviewer: It's a bit of a nit pick maybe, but would come in handy for navigation:
It would be helpful to hyperlink those labels, such as [accepted](https://github.com/opentofu/opentofu/issues?q=is%3Aopen+is%3Aissue+label%3Aaccepted), [pending-steering-committee-decision](https://github.com/opentofu/opentofu/issues?q=is%3Aopen+is%3Aissue+label%3Apending-steering-committee-decision) etc.

We currently only maintain a read-only mirror of these providers for the purposes of building binaries for the OpenTofu registry which would otherwise not be available. They are not true downstream versions that we can add patches to and OpenTofu users expect them to work the same as the upstream versions. In short: no, we cannot fix bugs or add features to these providers.

---

### Can you fork project X?

Currently, we are at capacity for development and do not have additional capacity to take on additional projects unless necessary for the continued work on OpenTofu. While the final determination lies with the [Technical Steering Committee](#who-is-the-technical-steering-committee), the answer is likely no in almost all cases.

---

### Who is the core team?

Core team members are full time developers sponsored by the participating companies. You can find the list of core developers in the [MAINTAINERS](MAINTAINERS) file. Their role is to triage issues, work on feature development, help plan, review, and document community contributions, support the community, and refer feature requests to the Technical Steering Committee.

---

### Can I become a core team member?

> Possibly, please look for open positions with the sponsoring companies as they hire core team members. The interview process is the same regardless of which sponsoring company you apply to. To become a core team member, you must be equally good at Go and at communication since much of our work is helping the community. Good luck!

Reviewer 6: I believe this should be "Possibly. Please look...".

Reviewer: It would be helpful to redirect people to a JOBS or Career page that lists the open positions from the sponsoring companies.

Author: Sorry, I was out of office. Unfortunately, I don't think all member companies have a dedicated careers page. :(

- [`accepted`](https://github.com/opentofu/opentofu/labels/accepted): the issue is accepted for development by either the core team or a community contributor. Please check if it's assigned to someone and comment on the issue if you want to work on it.
- [`help wanted`](https://github.com/opentofu/opentofu/labels/help%20wanted): this issue is open for community contributions. Please check if it's assigned to someone and comment on the issue if you want to work on it.
- [`good first issue`](https://github.com/opentofu/opentofu/labels/good%20first%20issue): this issue is relatively simple. If you are looking for a first contribution, this may be for you. Please check if it's assigned to someone and comment on the issue if you want to work on it.
- [`bug`](https://github.com/opentofu/opentofu/labels/bug): it's broken and needs to be fixed.
- [`enhancement`](https://github.com/opentofu/opentofu/labels/enhancement): it's a short-form feature request. If the implementation path is unclear, an `rfc` may be needed in addition.
- [`documentation`](https://github.com/opentofu/opentofu/labels/documentation): something that needs a description on the OpenTofu website.

> - [`rfc`](https://github.com/opentofu/opentofu/labels/rfc): a long-form feature request with details on how exactly it should be implemented.

Author: I'll try to get all issue labels updated once this is merged.
